rspotify-http/src/reqwest.rs

```diff
@@ -1,49 +1,23 @@
 //! The client implementation for the reqwest HTTP client, which is async by
 //! default.
 
-use super::{BaseHttpClient, Form, Headers, HttpError, HttpResult, Query};
+use super::{BaseHttpClient, Form, Headers, Query};
 
 use std::convert::TryInto;
 
 use maybe_async::async_impl;
-use reqwest::{Method, RequestBuilder, StatusCode};
-use rspotify_model::ApiError;
+use reqwest::{Method, RequestBuilder};
 use serde_json::Value;
 
-impl HttpError {
-    pub async fn from_response(response: reqwest::Response) -> Self {
-        match response.status() {
-            StatusCode::UNAUTHORIZED => Self::Unauthorized,
-            StatusCode::TOO_MANY_REQUESTS => Self::RateLimited(
-                response
-                    .headers()
-                    .get(reqwest::header::RETRY_AFTER)
-                    .and_then(|header| header.to_str().ok())
-                    .and_then(|duration| duration.parse().ok()),
-            ),
-            status @ StatusCode::FORBIDDEN | status @ StatusCode::NOT_FOUND => response
-                .json::<ApiError>()
-                .await
-                .map(Into::into)
-                .unwrap_or_else(|_| status.into()),
-            status => status.into(),
-        }
-    }
-}
-
-impl From<reqwest::Error> for HttpError {
-    fn from(err: reqwest::Error) -> Self {
-        Self::Request(err.to_string())
-    }
-}
+/// Custom enum that contains all the possible errors that may occur when using
+/// `reqwest`.
+#[derive(thiserror::Error, Debug)]
+pub enum ReqwestError {
+    #[error("request: {0}")]
+    Client(#[from] reqwest::Error),
```

Comment:
It's a little confusing that it seems `Client` and `StatusCode` are not under the same abstract level, is it fine to  wrap them into the same variant? And what does `Client` mean? What is the use case for `Client`?
Then, `StatusCode` doesn't sound like an error?

Reply:
The thing is that most HTTP libraries have two kinds of errors when making a request:

* An error in the client: when making the request (for example there's no internet or a DNS failure occurs), converting the response into UTF-8, etc
* An error in the server you made a request to: a 404, 503, 403, etc. Those codes considered as successful will not be included here (< 400).

It is fine to wrap them up under the same enum because both are errors after all.